### ISSUE TITLE
Squash commits when cloning

### ISSUE THREAD

**Issue by davbeek** [repobee/repobee#899]
Coming from GitHub Classroom, with [Template repositories](https://docs.github.com/en/github/creating-cloning-and-archiving-repositories/creating-a-template-repository), I am used to cloned repositories having a [single initial commit](https://docs.github.com/en/github/creating-cloning-and-archiving-repositories/creating-a-repository-from-a-template).

Repobee, however, leaves all commits intact. I think users would prefer to have all commits squashed into a single initial commit. To achieve this now, the user has to ensure that the template repos have a single commit, which is cumbersome. It would be nice if this could be done by Repobee. Something like:

1. Clone the template repo locally in Repobee
2. Delete the `.git` directory
3. `git init; git add .; git commit -m "Initial commit"`

Then clone using this local repo by means of `git push --set-upstream git@... master`

Could this be done in Repobee?

**Comment by rjglasse** [comment]
Have you checked out https://github.com/repobee/repobee-sanitizer?

We use it to prepare templates, using a solutions branch and a student branch. Sanitizer will strip away solutions based on a simple markup, and the result is a clean student branch with no commit history and just a single commit.

The developers of an assignment can still see the full commit history, but students will see no history once they get a repo 🙂

**Comment by davbeek** [comment]
I had indeed checked out everything in Repobee, including the sanitizer, but even now, after having reread the santizer documentation, I don't have a clue as to what to do in order to get the single commit you refer to.

**Comment by rjglasse** [comment]
We probably need a more clear example. Here's my workflow and I hope it clarifies.

1. Create a repo and create a `solutions` branch. The `master` branch will be the one that the student gets during `setup`
2. Checkout `solutions` and add all the material there as normal and commit/push the changes
3. Call sanitizer whilst on the `solutions` branch, and target the `master` branch for the single, sanitized commit
```
% repobee sanitize repo --target-branch master
hook results for sanitize repo

sanitize-repo: SUCCESS
Successfully sanitized repo
```
4. Now you checkout the `master` branch and you will get this feedback:
``` 
% git checkout master
Switched to branch 'master'
Your branch is ahead of 'origin/master' by 1 commit.
  (use "git push" to publish your local commits)
```

5. Simply push that commit and the `master` branch will reflect the changes, but hide their history.

You can repeat this process as many times as you like. Now all work is done on the `solutions` branch, and the `master` branch does not need to be touched (and probably should be locked). Let me know if it helps!

**Comment by davbeek** [comment]
That sound like rebase and squash the solutions branch onto master. I could do this without the sanitizer, or am I missing something?

**Comment by rjglasse** [comment]
Yes, I guess the bigger picture of the Sanitizer is to move us towards "solutions-driven development" of assignments, such that we have a nice warts-and-all history of the evolution of both the solution and assignment that live in one branch, but we can be safe in the knowledge that both the history and solution are hidden from the student when we distribute assignments. So perhaps a little overkill :-) I'm sure others will chime in with their thoughts as it could be a useful feature to consider.

**Comment by davbeek** [comment]
Thanks, clear! I personally would never want my solutions to live in the same repo as the one distributed to the students, taking [Murphy](https://en.wikipedia.org/wiki/Murphy%27s_law) into consideration. In fact, I know from a student from a course of exactly such an error, where a solution was unintentionally distributed together with an assignment.

Also, I am looking for a solution that is as simple and straightforward as possible, for a user base without a computer science background. The idea of squashing when cloning is that the user/teacher, who is already overloaded with too many things to do, does not need to do _any additional_ operation to get the desired result.

**Comment by rjglasse** [comment]
Before using sanitizer in my workflow, I would have agreed with you completely. Having used it, my position is shifting. On the surface it avoids the chance of drift between solution and assignment that can happen over time, whilst also reducing the effort in maintaining two separate repositories for essentially the same thing. More fundamentally, much like test-driven development, it shifts your mindset when creating an assignment, and introduces an explicit step to my workflow, much like using a commit to document change, that this assignment is ready to distribute. That all said, humans still make mistakes (and we have no good fix for that) ;-)

In any case, it's really interesting to hear other teacher's experiences and needs re: RepoBee and I agree it should  minimise their labour be as simple as possible to use.

**Comment by slarse** [comment]
@davbeek I'm not quite sure what you mean by "squash on clone". As you mention cloning template repositories, my understanding is that on running `repos setup`, you want student repos setup with a single, squashed commit of the templates. Is this correct?

In that case, the reason RepoBee does not squash the template repo commits before setting up the student repos is that it would make pushing updates without force pushing impossible. And we sometimes need to push updates, because we mess up with the assignments and realize only after publishing.

It would however be pretty easy to add squashing of template repos, but that would have to be taken with the drawback that `repos update` would be very much hampered in doing its job.

> That sounds like rebase and squash the solutions branch onto master.

It's kind of like that, but Sanitizer uses lower-level Git commands such that merge conflicts become impossible. In fact, it doesn't take the state of the master branch into account at all, it just creates a new commit on it based on the sanitized solutions branch (i.e. after removing solutions).

The documentation still isn't quite up-to-snuff, we're working on that. But Sanitizer is all about maintaining solutions and assignments, and this isn't what you're looking for, is it? You just want the student repos to start out with a single, squashed commit, right?

**Comment by davbeek** [comment]
> you want student repos setup with a single, squashed commit of the templates. Is this correct?

Yes indeed, and I should have stated this as _Squash commits when setting up student repos_. I got the terminology mixed up.

> the reason RepoBee does not squash the template repo commits before setting up the student repos is that it would make pushing updates without force pushing impossible

Interesting! I didn't realize there was a trade-off involved.

> But Sanitizer is all about maintaining solutions and assignments, and this isn't what you're looking for, is it?

No, indeed it isn't.

> You just want the student repos to start out with a single, squashed commit, right?

Yes, but I didn't know about the problem of then not being able to push updates. I fortunately never have been in that situation and I would have to consider whether I ever might want to push an update. Such an update might then also interfere with work already done by students?

**Comment by davbeek** [comment]
Just a wild thought: could you perhaps first create a kind of "master repo" with a squashed commit on the target organization and then "kind of fork" the student repos from that "master repo", so that in the special case that one would need to push an update, the update could be done to the master repo? Perhaps that would be to complicated, but it would help me in understanding the problem.

**Comment by slarse** [comment]
> I fortunately never have been in that situation and I would have to consider whether I ever might want to push an update. Such an update might then also interfere with work already done by students?

The way update works is that it just tries to push the new state of the template repository to each student repository, without force pushing. If the push is rejected (because the students have already pushed something), then a customizable issue is opened in the student's repository instead, which would typically contain the patch the student should apply and any additional instructions required.

The problem then with squashing is that, as squash commits are unique each time you do them (timestamps etc change), every push would be rejected even when pushing the _same_ version of the template repo. To work around this, we would have to check the hash of the root Git tree and compare that to the version of the template repository that was last used to push to the student repositories. But how do we know which was the last version used? You can see here that it gets rather complicated, because we need to go outside of the features natively supported by Git.

> Just a wild thought: could you perhaps first create a kind of "master repo" with a squashed commit on the target organization and then "kind of fork"

This touches closely to an idea I had before, where I wanted to enable on-the-fly preprocessing of student repositories before publishing. I later came to the realization that no one would dare to use that kind of feature because you really want to inspect the state of the repository before publishing to students, and so scrapped it.

In any case, this suffered from the same problem as squashing, as each time one runs preprocessing and commits it, the commit is unique by timestamp. The idea to solve it was to create a "target organization-local" template repository, which would essentially cache the on-the-fly-updates. When running setup, this repo is created if it does not already exist. When running update, this repository would be used as a ground truth for the current starter state of the student repos. When pushing an update, RepoBee would clone that target organization-local repository, make a merge-conflict free squash merge with it (like Sanitizer does), and then push _that_ one instead.

While I have no problem making squashing available (it's really not hard to implement), I'm a bit hesitant making it "update compatible" with target organization-local templates without it being heavily requested, as it is a bit hard and would need to be very well tested.

That being said, would you like squashing to be implemented, with update support potentially (but not guaranteed) coming in the future?

**Comment by davbeek** [comment]
Thanks @slarse for your elaborate reply. I like your "target organization-local" template repository, which I think, is equivalent to my "master repo", which is created when setting up the student repos and which contains a single commit, equivalent to the squashed version of the user template repo.

When updates are needed, I would require the user to do them exclusively in this target organization-local repo and I would not apply any squashing on the update commits in this repo. So effectively, all Repobee would need to do in the case of presence of this target organization-local template repo is to use this repo for any future updates instead of the repo from which it originated (the unsquashed version).

The user would then be able to choose between two options:

- The current Repobee approach of not applying any squashing, which is very user friendly in the case updates are needed after having set up the student repos.
- One initial squash, leading to the creation of a target org-local repo having a single initial commit. Any updates would have to be committed by the user to this target org-local repo, so that the updates can be pushed (without squashing) to the student repos by Repobee.

> That being said, would you like squashing to be implemented, with update support potentially (but not guaranteed) coming in the future?

Yes! I have zero experience with updates after having setup student repos and I would currently be perfectly OK with a solution in which squashing is offered by Repobee at the expense of pushing updates to student repos.

I would like to add that I completely agree with your hesitancy to implement complex features.

**Comment by slarse** [comment]
> When updates are needed, I would require the user to do them exclusively in this target organization-local repo and I would not apply any squashing on the update commits in this repo. So effectively, all Repobee would need to do in the case of presence of this target organization-local template repo is to use this repo for any future updates instead of the repo from which it originated (the unsquashed version).

Just a note on this model; we used to have all our template repos in the target organization in a course with 20+ assignments (one template for each), with solutions in separate repositories, and doing updates in the target organization templates. It was all but impossible to keep everything synchronized, leading to the inception of the template organization as well as Sanitizer.

I'm not knocking on your model, if it works for you then that's great :). But it proved too much of a hassle for us and our coworkers, which is why we strongly recommend using a dedicated template organization for templates, with just one ground truth for each template.

> Yes! I have zero experience with updates after having setup student repos and I would currently be perfectly OK with a solution in which squashing is offered by Repobee at the expense of pushing updates to student repos.

Let's start with this then, and then we'll see what we do from there!

@tohanss Care to give it a go? What we need here is:

1. `repos setup` must take an additional flag `--squash`, which if specified should cause the template repository to be squashed before being pushed to students.
2. Actually, that's it.

**Comment by davbeek** [comment]
> which is why we strongly recommend using a dedicated template organization for templates, with just one ground truth for each template.

Agree. In fact, we do have a separate template organization for our (four) templates. These templates, however, tend to get multiple commits over time. The idea would be to let the squashed version of each template be dynamically created on the target organization as a result of the `--squash` flag.

**Comment by slarse** [comment]
> The idea would be to let the squashed version of each template be dynamically created on the target organization as a result of the --squash flag, when setting up the student repos.

Sorry for the late response, this notification blew past me!

Yeah, I see that use case. Technically, that's should already be possible by doing something like this (assuming the `template_org_name` is _not_ configured but `org_name` is).

```bash
$ git clone <base-url>/some-namespace/some-template-org.git --depth 1 # fetch only last commit
$ repobee repos migrate -a some-template-org # migrate template into target org
# now use RepoBee as per usual
```

Fetching with depth 1 doesn't squash per-se as it retains the commit message, but for all intents and purposes it removes the commit message.

Of course, having this as a `--squash` flag is simpler, as it lets one have the template org configured as usual.

**Comment by davbeek** [comment]
Thanks, interesting solution with three problems:
- The commit message that the students will see makes no sense, because it refers to the last commit only.
- GitKraken does not open the cloned repo. Obviously this is a GitKraken bug, because git, Sourcetree and SmartGit all work. Still all my students use GitKraken.
- This requires some git knowledge, which is what I would like to avoid as much as possible: my target group is teachers who just want to get things done, as quickly and simply as possible.

**Comment by slarse** [comment]
Yeah, not a perfect solution by any means. Didn't know it would upset the kraken, though, but I'd imagine that users of Git GUI apps aren't particularly prone to do "strange stuff".

In any case, @tohans is going to try to add a `--squash` flag to the `setup` command. Perhaps it would also be desirable to supply a commit message with that? Something like `--squash "Add project"`.

**Comment by davbeek** [comment]
I can imagine an optional commit message, but I would also be fine with the default git initial commit message: "Initial commit".

**Comment by slarse** [comment]
> but I would also be fine with the default git initial commit message: "Initial commit".

Actually, that seems like a better idea to me. It's a reasonable squash message that I think very few would want to change.

**Comment by davbeek** [comment]
> It's a reasonable squash message that I think very few would want to change.

Yes!
